File: src/Runtime/Optimized/wasm_src/src/mmd_runtime_bone.rs
use nalgebra::{Vector3, UnitQuaternion, Matrix4 };

use crate::{animation_arena::AnimationArena, append_transform_solver::AppendTransformSolverArena};

pub(crate) struct MmdRuntimeBoneArena {
    arena: Box<[MmdRuntimeBone]>,
    bone_stack: Vec<usize>,
}

impl MmdRuntimeBoneArena {
    pub fn new(area: Box<[MmdRuntimeBone]>, bone_stack: Vec<usize>) -> Self {
        MmdRuntimeBoneArena {
            arena: area,
            bone_stack,
        }
    }

    pub fn update_world_matrix(&mut self, root: usize) {
        let stack = &mut self.bone_stack;
        stack.push(root);

        while let Some(bone) = stack.pop() {
            if let Some(parent_bone) = self.arena[bone].parent_bone {
                let parent_world_matrix = self.arena[parent_bone].world_matrix;

                let bone = &mut self.arena[bone];
                bone.world_matrix = parent_world_matrix * bone.local_matrix;
            } else {
                let bone = &mut self.arena[bone];
                bone.world_matrix = bone.local_matrix;
            }

            let bone = &self.arena[bone];
            for child_bone in &bone.child_bones {
                stack.push(*child_bone);
            }
        }
    }
}

impl std::ops::Deref for MmdRuntimeBoneArena {
    type Target = [MmdRuntimeBone];

    fn deref(&self) -> &Self::Target {
        &self.arena
    }
}

impl std::ops::DerefMut for MmdRuntimeBoneArena {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.arena
    }
}

pub(crate) struct MmdRuntimeBone {
    pub rest_position: Vector3<f32>,
    index: usize,

    pub parent_bone: Option<usize>,
    pub child_bones: Vec<usize>,
    pub transform_order: i32,
    pub transform_after_physics: bool,

    pub append_transform_solver: Option<usize>,
    pub ik_solver: Option<usize>,

    pub morph_position_offset: Option<Vector3<f32>>,
    pub morph_rotation_offset: Option<UnitQuaternion<f32>>,

    pub ik_rotation: Option<UnitQuaternion<f32>>,

    pub local_matrix: Matrix4<f32>,
    pub world_matrix: Matrix4<f32>,
}

impl MmdRuntimeBone {
    pub fn new(index: usize) -> Self {
        MmdRuntimeBone {
            rest_position: Vector3::zeros(),
            index,
            
            parent_bone: None,
            child_bones: Vec::new(),
            transform_order: 0,
            transform_after_physics: false,

            append_transform_solver: None,
            ik_solver: None,

            morph_position_offset: None,
            morph_rotation_offset: None,

            ik_rotation: None,

            local_matrix: Matrix4::identity(),
            world_matrix: Matrix4::identity(),
        }
    }

    pub fn animated_position(&self, animation_arena: &AnimationArena) -> Vector3<f32> {
        let mut position = animation_arena.nth_bone_position(self.index);
        if let Some(morph_position_offset) = self.morph_position_offset {
            position += morph_position_offset;
        }
        position
    }

    pub fn animated_rotation(&self, animation_arena: &AnimationArena) -> UnitQuaternion<f32> {
        let mut rotation = animation_arena.nth_bone_rotation(self.index);
        if let Some(morph_rotation_offset) = self.morph_rotation_offset {
            rotation *= morph_rotation_offset;
        }
        rotation
    }

    pub fn animation_position_offset(&self) -> Vector3<f32> {
        let mut position = Vector3::zeros();
        if let Some(morph_position_offset) = self.morph_position_offset {
            position += morph_position_offset;
        }
        position - self.rest_position
    }

    pub fn update_local_matrix(&mut self, animation_arena: &AnimationArena, append_transform_solver_arena: &AppendTransformSolverArena) {
        let mut rotation = self.animated_rotation(animation_arena);
        if let Some(ik_rotation) = self.ik_rotation {
            rotation = ik_rotation * rotation;
        }

        let mut position = self.animated_position(animation_arena);
        
        if let Some(append_transform_solver) = self.append_transform_solver {
            let append_transform_solver = &append_transform_solver_arena[append_transform_solver];

            if append_transform_solver.is_affect_rotation() {
                rotation *= append_transform_solver.append_rotation_offset();
            }
            if append_transform_solver.is_affect_position() {
                position += append_transform_solver.append_position_offset();
            }
        }

        self.local_matrix = 
            Matrix4::new_translation(&position) *
            rotation.to_homogeneous() *
            Matrix4::new_nonuniform_scaling(&animation_arena.nth_bone_scale(self.index));
    }
}

File: src/Runtime/mmdRuntimeBone.ts
import { Matrix, Quaternion, Vector3 } from "@babylonjs/core/Maths/math.vector";
import type { Nullable } from "@babylonjs/core/types";

import type { MmdModelMetadata } from "@/Loader/mmdModelMetadata";
import { PmxObject } from "@/Loader/Parser/pmxObject";

import type { AppendTransformSolver } from "./appendTransformSolver";
import type { IkSolver } from "./ikSolver";
import type { IMmdRuntimeLinkedBone } from "./IMmdRuntimeLinkedBone";

/**
 * Bone for MMD runtime
 *
 * For mmd runtime, it is necessary to override the bone system because it has a different implementation than the usual matrix update method
 *
 * Which requires the mmd runtime bone, which is the wrapper of the babylon.js bone
 */
export interface IMmdRuntimeBone {
    /**
     * The Babylon.js bone
     */
    readonly linkedBone: IMmdRuntimeLinkedBone;

    /**
     * Name of the bone
     */
    readonly name: string;

    /**
     * Parent bone
     */
    readonly parentBone: Nullable<IMmdRuntimeBone>;

    /**
     * Child bones
     */
    readonly childBones: readonly IMmdRuntimeBone[];

    /**
     * Transform order
     */
    readonly transformOrder: number;

    /**
     * Bone flag
     *
     * @see PmxObject.Bone.Flag
     */
    readonly flag: number;

    /**
     * Whether the bone transform is applied after physics
     */
    readonly transformAfterPhysics: boolean;

    /**
     * Get ik solver index
     *
     * If the bone does not have an ik solver, it will return -1
     */
    get ikSolverIndex(): number;
}

export class MmdRuntimeBone implements IMmdRuntimeBone {
    public readonly linkedBone: IMmdRuntimeLinkedBone;

    public readonly name: string;
    public parentBone: Nullable<MmdRuntimeBone>;
    public readonly childBones: MmdRuntimeBone[];

    public readonly transformOrder: number;
    public readonly flag: number;
    public readonly transformAfterPhysics: boolean;

    public appendTransformSolver: Nullable<AppendTransformSolver>;
    public ikSolver: Nullable<IkSolver>;

    public readonly morphPositionOffset: Vector3;
    public readonly morphRotationOffset: Quaternion;

    public ikRotation: Nullable<Quaternion>;

    public readonly localMatrix: Matrix;
    public readonly worldMatrix: Matrix;

    public getAnimatedPositionToRef: (target: Vector3) => Vector3;
    public getAnimatedRotationToRef: (target: Quaternion) => Quaternion;
    public getAnimationPositionOffsetToRef: (target: Vector3) => Vector3;
    // public getAnimationRotationOffsetToRef: (target: Quaternion) => Quaternion;

    public constructor(babylonBone: IMmdRuntimeLinkedBone, boneMetadata: MmdModelMetadata.Bone) {
        this.linkedBone = babylonBone;

        this.name = boneMetadata.name;
        this.parentBone = null;
        this.childBones = [];

        this.transformOrder = boneMetadata.transformOrder;
        this.flag = boneMetadata.flag;
        this.transformAfterPhysics = (boneMetadata.flag & PmxObject.Bone.Flag.TransformAfterPhysics) !== 0;

        this.appendTransformSolver = null;
        this.ikSolver = null;

        this.morphPositionOffset = Vector3.Zero();
        this.morphRotationOffset = Quaternion.Identity();

        this.ikRotation = null;

        this.localMatrix = Matrix.Identity();
        this.worldMatrix = babylonBone.getFinalMatrix();

        this.getAnimatedPositionToRef = this._getAnimatedPositionToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationToRef;
        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetToRef;
    }

    private _getAnimatedPositionWithMorphToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        return target.addInPlace(this.morphPositionOffset);
    }

    private _getAnimatedPositionToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        return target;
    }

    private _getAnimatedRotationToRef(target: Quaternion): Quaternion {
        return target.copyFrom(this.linkedBone.rotationQuaternion);
    }

    private _getAnimatedRotationWithMorphToRef(target: Quaternion): Quaternion {
        target.copyFrom(this.linkedBone.rotationQuaternion);
        return target.multiplyInPlace(this.morphRotationOffset);
    }

    private static readonly _TempVector3 = new Vector3();

    private _getAnimationPositionOffsetToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempVector3);
        return target.subtractInPlace(MmdRuntimeBone._TempVector3);
    }

    private _getAnimationPositionOffsetWithMorphToRef(target: Vector3): Vector3 {
        target.copyFrom(this.linkedBone.position);
        target.addInPlace(this.morphPositionOffset);
        this.linkedBone.getRestMatrix().getTranslationToRef(MmdRuntimeBone._TempVector3);
        return target.subtractInPlace(MmdRuntimeBone._TempVector3);
    }

    // a: rest quaternion
    // b: animation quaternion
    // c: animated quaternion

    // a * b = c

    // to get b from a and c:
    // a^-1 * c = b

    // private static readonly _TempQuaternion = new Quaternion();

    // private _getAnimationRotationOffsetToRef(target: Quaternion): Quaternion {
    //     target.copyFrom(this.babylonBone.rotationQuaternion);
    //     Quaternion.FromRotationMatrixToRef(this.babylonBone.getRestMatrix(), MmdRuntimeBone._TempQuaternion).invertInPlace();
    //     return MmdRuntimeBone._TempQuaternion.multiplyInPlace(target);
    // }

    // private _getAnimationRotationOffsetWithMorphToRef(target: Quaternion): Quaternion {
    //     target.copyFrom(this.babylonBone.rotationQuaternion);
    //     target.multiplyInPlace(this.morphRotationOffset);
    //     Quaternion.FromRotationMatrixToRef(this.babylonBone.getRestMatrix(), MmdRuntimeBone._TempQuaternion).invertInPlace();
    //     return MmdRuntimeBone._TempQuaternion.multiplyInPlace(target);
    // }

    public enableMorph(): void {
        this.getAnimatedPositionToRef = this._getAnimatedPositionWithMorphToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationWithMorphToRef;

        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetWithMorphToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetWithMorphToRef;
    }

    public disableMorph(): void {
        this.getAnimatedPositionToRef = this._getAnimatedPositionToRef;
        this.getAnimatedRotationToRef = this._getAnimatedRotationToRef;

        this.getAnimationPositionOffsetToRef = this._getAnimationPositionOffsetToRef;
        // this.getAnimationRotationOffsetToRef = this._getAnimationRotationOffsetToRef;
    }

    private static readonly _TempRotation = Quaternion.Identity();
    private static readonly _TempPosition = Vector3.Zero();

    public updateLocalMatrix(): void {
        const rotation = this.getAnimatedRotationToRef(MmdRuntimeBone._TempRotation);
        if (this.ikRotation !== null) {
            this.ikRotation.multiplyToRef(rotation, rotation);
        }

        const position = this.getAnimatedPositionToRef(MmdRuntimeBone._TempPosition);

        if (this.appendTransformSolver !== null) {
            if (this.appendTransformSolver.affectRotation) {
                rotation.multiplyInPlace(this.appendTransformSolver.appendRotationOffset);
            }
            if (this.appendTransformSolver.affectPosition) {
                position.addInPlace(this.appendTransformSolver.appendPositionOffset);
            }
        }

        Matrix.ComposeToRef(
            this.linkedBone.scaling,
            rotation,
            position,
            this.localMatrix
        );
    }

    private static readonly _Stack: MmdRuntimeBone[] = [];

    public updateWorldMatrix(): void {
        const stack = MmdRuntimeBone._Stack;
        stack.length = 0;
        stack.push(this);

        while (stack.length > 0) {
            const bone = stack.pop()!;

            const parentBone = bone.parentBone;
            if (parentBone !== null) {
                bone.localMatrix.multiplyToRef(parentBone.worldMatrix, bone.worldMatrix);
            } else {
                bone.worldMatrix.copyFrom(bone.localMatrix);
            }

            const childrenBones = bone.childBones;
            for (let i = 0, l = childrenBones.length; i < l; ++i) {
                stack.push(childrenBones[i]);
            }
        }
    }

    /**
     * Get ik solver index
     *
     * If the bone does not have an ik solver, it will return -1
     */
    public get ikSolverIndex(): number {
        return this.ikSolver !== null ? this.ikSolver.index : -1;
    }
}
